themes/src/main/resources/theme/keycloak.v2/login/resources/js/password-policy.js

```diff
@@ -0,0 +1,56 @@
+const policies = {
+  length: (policy, value) => {
+    if (value.length < policy.value) {
+      return templateError(policy);
+    }
+  },
+  maxLength: (policy, value) => {
+    if (value.length > policy.value) {
+      return templateError(policy);
+    }
+  },
+  upperCase: (policy, value) => {
+    if (
+      value.split("").filter((char) => char !== char.toUpperCase()).length >
+      policy.value
+    ) {
+      return templateError(policy);
+    }
+  },
+  lowerCase: (policy, value) => {
+    if (
+      value.split("").filter((char) => char !== char.toLowerCase()).length >
+      policy.value
+    ) {
+      return templateError(policy);
+    }
+  },
+  digits: (policy, value) => {
+    const digits = value.split("").filter((char) => char.match(/\d/));
+    if (digits.length < policy.value) {
+      return templateError(policy);
+    }
+  },
+  specialChars: (policy, value) => {
+    let specialChars = value.split("").filter((char) => char.match(/\W/));
+    if (specialChars.length < policy.value) {
+      return templateError(policy);
+    }
+  },
+};
+
+const templateError = (policy) => policy.error.replace("{0}", policy.value);
+
+export function validatePassword(password, activePolicies) {
+  const errors = [];
+  const policiesNames = activePolicies.map((policy) => Object.keys(policy)[0]);
+  for (let i = 0; i < policiesNames.length; i++) {
+    const policyName = policiesNames[i];
+    let policy = policies[policyName];
+    let validationError = policy(activePolicies[i][policyName], password);
+    if (validationError) {
+      errors.push(validationError);
```

Comment:
This seems like it could be a `.map().filter()` operation. It also seems the `i` variable is never used, so this could be a regular [`for..of`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of) loop.

Reply:
`i` is used

Reply:
Missed that one, my bad.